Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
     m_innerLegacyAttachment->m_thumbnail = WTFMove(m_thumbnail);
     m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
     m_innerLegacyAttachment->m_iconSize = m_iconSize;
-    m_innerLegacyAttachment->setIdAttribute(attachmentPreviewIdentifier());
-    container->appendChild(*m_innerLegacyAttachment);
+    addShadowElement(m_innerLegacyAttachment, attachmentPreviewIdentifier(), String { });
+
+    m_elementWithAction = createShadowDivElement(attachmentActionIdentifier(), attachmentActionForDisplay());
+
+    m_elementWithTitle = createShadowDivElement(attachmentTitleIdentifier(), attachmentTitleForDisplay());
+
+    m_elementWithSubtitle = createShadowDivElement(attachmentSubtitleIdentifier(), attachmentSubtitleForDisplay());
+
+    updateSave(attributeWithoutSynchronization(saveQualifiedName()));
+}
 
-    m_elementWithAction = HTMLDivElement::create(document());
-    m_elementWithAction->setIdAttribute(attachmentActionIdentifier());
-    if (const auto& action = attachmentActionForDisplay(); !action.isEmpty())
-        m_elementWithAction->setInnerText(String { action });
-    container->appendChild(*m_elementWithAction);
+class AttachmentSaveEventListener final : public EventListener {
+public:
+    static Ref<AttachmentSaveEventListener> create(HTMLAttachmentElement& attachment) { return adoptRef(*new AttachmentSaveEventListener(attachment)); }
 
-    m_elementWithTitle = HTMLDivElement::create(document());
-    m_elementWithTitle->setIdAttribute(attachmentTitleIdentifier());
-    if (auto title = attachmentTitleForDisplay(); !title.isEmpty())
-        m_elementWithTitle->setInnerText(WTFMove(title));
-    container->appendChild(*m_elementWithTitle);
+    bool operator==(const EventListener& other) const final
+    {
+        return this == &other;
+    }
 
-    m_elementWithSubtitle = HTMLDivElement::create(document());
-    m_elementWithSubtitle->setIdAttribute(attachmentSubtitleIdentifier());
-    if (auto subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
-        m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    void handleEvent(ScriptExecutionContext&, Event& event) final
+    {
+        if (event.type() == eventNames().clickEvent) {
+            event.preventDefault();
+            event.stopPropagation();
+            event.stopImmediatePropagation();
+
+            CustomEvent::Init init;
+            init.bubbles = true;
+            init.cancelable = true;
+            init.composed = true;
+            m_attachment->dispatchEvent(CustomEvent::create(m_attachment->attributeWithoutSynchronization(saveQualifiedName()), init, event.isTrusted() ? Event::IsTrusted::Yes : Event::IsTrusted::No));
```

Comment:
Feels kinda weird that this just emits a custom event that has what's inside the save attribute, but I assume you coordinated with other teams using `<attachment>` and ended up with this solution.

Reply:
I didn't coordinate with the Mail team yet, but they're going to start using it soon, so I'll hopefully get feedback about that!
Since this element is inside a user agent shadow DOM, it's not reachable by their code, so they can't attach a listener directly to it; and I can't just pass a standard `click` event because there would be no way to distinguish it from a click on the whole attachment brick -- but please let me know if I'm incorrect.
And I wanted to give our client a way to potentially distinguish Save-clicks from different attachments (just in case), that's why I'm using the `save` attribute value to name the custom event.

What else would you recommend, that would feel less weird?

Reply:
It purely depends on what the client needs. From a web developer POV, I would find a single event name easier to handle, with the detail of what's being saved in the event detail/data, simply because I would be able to watch them from a single event listener.


Although if there are no specific needs, a `<a download>` tag pointing to a blob might just be fine as well.

Reply:
If this were JS, I'd do: `new CustomEvent("save", { detail: [...] });`

Reply:
Thanks for the suggestion. I'll iterate with the client devs, see if they'd like this or something else...